engine/gamesys/src/gamesys/components/comp_model.cpp

```diff
@@ -280,46 +280,56 @@ namespace dmGameSystem
         return GetMaterialResource(component, resource, index)->m_Material;
     }
 
-    TextureResource* GetTextureResource(const ModelComponent* component, uint32_t material_index, uint32_t texture_unit)
+    static TextureResource* GetTextureFromSamplerNameHash(const MaterialInfo* material_info, const MaterialResource* material, uint32_t material_texture_index, dmhash_t sampler_name_hash)
     {
-        assert(texture_unit < MAX_TEXTURE_COUNT);
-        TextureResource* texture;
-
-        // Overridden textures (from Lua code)
-        texture = component->m_Textures[texture_unit];
-        if (texture)
-            return texture;
-        // Overridden material (from Lua code)
-        MaterialResource* material = component->m_Material;
-        if (material)
-            texture = material->m_Textures[texture_unit];
-        if (texture)
-            return texture;
+        // Material is the actual selected material (may be overridden at runtime)
 
-        MaterialInfo* material_info = &component->m_Resource->m_Materials[material_index];
-        // Overridden textures (from .model)
-        MaterialTextureInfo* texture_infos = material_info->m_Textures;
-        if (material_info->m_Textures && texture_unit < material_info->m_TexturesCount)
-            texture = material_info->m_Textures[texture_unit].m_Texture;
-        if (texture)
-            return texture;
-
-        // The textures which are set in the .material file
-        material = material_info->m_Material;
-        if (material)
-            texture = material->m_Textures[texture_unit];
-        if (texture)
-            return texture;
+        // If the model has textures, let's find the one associated with the sampler name
+        for (uint32_t i = 0; i < material_info->m_TexturesCount; ++i)
+        {
+            if (material_info->m_Textures[i].m_SamplerNameHash == sampler_name_hash)
+                return material_info->m_Textures[i].m_Texture;
+        }
+
+        // If it is contains materials, use that
+        if (material_texture_index < material->m_NumTextures)
+            return material->m_Textures[material_texture_index];
 
         return 0;
     }
 
-    dmGraphics::HTexture GetMaterialTexture(const ModelComponent* component, uint32_t material_index, uint32_t texture_unit)
+    // A bit legacy, as we don't want to set textures based on index anymore, but sampler names.
+    static TextureResource* GetTextureResource(const ModelComponent* component, uint32_t material_index, uint32_t material_texture_index)
```

Comment:
This is for our Lua api, to get a texture by index, which makes less sense nowadays, but is still part of the lua api.